File: Top Interview Questions/src/14.longest-common-prefix.py
#
# @lc app=leetcode id=14 lang=python3
#
# [14] Longest Common Prefix
#
from typing import List
# @lc code=start
class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        if len(strs) > 0:
            prefix = strs[0]
            for i in range(1, len(strs)):
                each = strs[i]

                while True:
                    if each[:len(prefix)] == prefix:
                        break
                    elif prefix == "":
                        return prefix
                    prefix = prefix[:-1]                        

            return prefix
        else:
            return ""

# @lc code=end
s = Solution()
print(s.longestCommonPrefix(["flower","flow","flight"]))

File: Top Interview Questions/src/14.longest-common-prefix.java
/*
 * @lc app=leetcode id=14 lang=java
 *
 * [14] Longest Common Prefix
 */

// @lc code=start
class Solution {
  public String longestCommonPrefix(String[] strs) {
    if (strs.length == 1) {
      return strs[0];
    }

    StringBuilder prefix = new StringBuilder();

    // 첫 번째 글자를 기준으로 prefix 생성
    for (int i = 0; i < strs[0].length(); i++) {
      prefix.append(strs[0].charAt(i));
      
      for (String each : strs) {
        if (each.indexOf(prefix.toString()) == 0) {
          ;
        } else {
          // prefix가 아닌 게 발견되면, 바로 이전 prefix까지가 정답임
          prefix.setLength(prefix.length() - 1);
          return prefix.toString();
        }
      }
    }

    return prefix.toString();
  }
}
// @lc code=end

